docs/documentation/namespaces-proposal.md

```diff
@@ -0,0 +1,93 @@
+# Namespaces Proposal
+
+This document contains a proposal for a guideline on how to expand the usage of 
+php namespaces to ILIAS modules and services. The suggested approach will elaborate 
+the following aspects:
+* naming conventions
+* integration of composer and autoloading
+* handling of subdirectories and interfaces
+* migration of the current structure
+* possible problems
+
+## Guideline
+
+### Naming & Structuring
+
+* The naming and structuring of all Namespaces MUST follow the PHP Standards Recommendation PSR-4: https://www.php-fig.org/psr/psr-4/.
+* The base prefix namespace is *ILIAS* and is pointing at the subdirectory */src*. Therefore, all Modules and Services MUST be located in the directory */src*. 
+* The subdirectory */classes* for Modules and Services will not be needed anymore and MUST be removed.
+* The prefixes "*class.*" and "*il*" for ILIAS classes will not be needed anymore and MUST be removed.
+* Newly created interfaces, abstract classes and traits MUST follow the PSR Naming Conventions: https://www.php-fig.org/bylaws/psr-naming-conventions. Already existing interfaces, abstract classes and traits SHOULD be renamed.
```

Comment:
The `PSR Naming conventions` are the convention of the PHP Framework Interop Group (php-fig). It is used for proposals and the defined standards by this group. So `PSR Naming conventions` are not suitable for this project(e.g. vendor namespace `Psr`). Maybe we should extract the relevant rules from the PHP Fig document into this document instead.

Reply:
I like to underline a point made by @legionth, your cited conventions are "by-laws" of PHP Standards Recommendations from the PHP-FIG group. Certainly, we should look at them and cherry pick the pearls that fit for us. The complete set however is not productive. I would hate to loss class. just to get *Interface.

Reply:
I understand the point, but I think we should not jump to conclusions here.

Naming implementations and interfaces differently is not only a convention of the PHP Framework Interop Group, but this is also a common practive in many of the libs that we are using (twig, monolog, pimple, whoops, ...). Others go with an "I" prefix (phpoffice, sabre/dav).

If our main goal is to attract new developers by adopting common rules and best practices you should take into consideration to go with this or a very similar naming convention.

Additionally I think that it would be good to be able to tell if something is an interface or not by just looking at the code.

E.g. given an API function (or constructor doing DI):
```
// here using the I prefix for interfaces
function ...(IThingA $a, IThingB, $b, ThingC $c): IThingD
```
You will question immediately why ThingC is not an interface.

We will use interfaces for the definition of contracts between our components and for easing DI / testability, right? But not all code classes will implement interfaces in the future. I think it would help to see if something belongs to the "contract" part or not just by looking at its name in the code.

And last but not least it would prevent to force us to use either aliases (which need another naming convetion) or qualified names to often, since there seems to be a tendency to an inflation of equally named pairs of classes/interfaces in the code otherwise.

Reply:
The discussion here gets highly subjective and I have the feeling it is losing the focus of my first recommendation.

As @Amstutz mentioned, I originally suggested to add the best parts of the `PSR Naming conventions`, because these were originally designed for the PHP FIG.
If these are useful or not can be discussed in an later discussion when they are really suggested by this PR. Currently there is only a link in this PR that can't be used for this project, so IMO it is pointless to discuss any further, without a solid base to discuss.

I didn't want to discuss of how classes and interfaces are named until there is an recommendation by @theodortruffer in this PR. Otherwise we might move into the wrong direction and too many discussions at once.

There are 8 points on the list of `PSR Naming conventions` and everyone is worth a discussion.

What do you think @theodortruffer picks the best parts and add the to the document, before we discuss any further? 

Naming conventions are a highly sensitive topic for developers and engineers. I think we need a step to step discussion, even if this needs a while to be completely settled.